### ISSUE TITLE
Voting/Presence CRDT

### ISSUE THREAD

**Issue by jroper** [akka/akka#27375]
I don't know if there's a generalised name or more general form for this CRDT, but the use case I have in mind is for online presence tracking, as is done when a social application, eg chat, tells you which of your friends are currently online.

The CRDT uses voting, if any node votes that the user is online, the user is online. So the resulting value of the CRDT is a boolean.

Online might be defined by having one or more active connections (eg websocket, gRPC stream, whatever) to a node. It is assumed that as long as a node is up, that node is able to correctly generate events when a user connects or disconnects, such that it will always ensure that its local representation of the CRDT has the correct value for its vote.

The vote itself can be represented using an integer - odd is true, even is false, the merge function always takes the highest value, so changing status is done by incrementing by one. The actual data, merge function, delta support is identical to a GCounter.

The final interesting piece of information is how the value is read - to read it, you must supply the current cluster membership state. Votes from members that are no longer part of the cluster are ignored - since a user can't be connected to node that isn't running. So, if a node crashes or is indefinitely partitioned, the votes it contributes cease to have any effect once it is downed.

There are some variants, such as a structure similar to PNCounter for tracking how many active connections a user has across the cluster, or even a Set structure, where each node can contribute items to it and nodes can only remove the items they add.

Could this be a generally useful CRDT that Akka distributed data provides?

**Comment by chbatey** [comment]
Maybe we could start with a sample to show how to compose it from existing CRDTs and then decide if it is general enough to be included?

**Comment by jroper** [comment]
I don't think it is composable from existing CRDTs - it may have the same state as a `GCounter`, but the state of the `GCounter` is `private [akka]`, there's no way to inspect each nodes counter value. `GCounter` only exposes the summed values of all the nodes as a public value. So the only way to implement it outside of the Akka codebase is from scratch.

**Comment by helena** [comment]
@jroper Do you mean the [boolean Flag](https://github.com/akka/akka/blob/master/akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala)?

**Comment by jroper** [comment]
@helena Flag starts set to false, and then can only be set to true, once set to true, it can't be set back to false. That's not useful for indicating whether a user is currently connected to the system or not.

**Comment by patriknw** [comment]
I’m positive to some specialized data type for this use case. Items that are ”owned” by a node is very useful, I have used that for other things such as cluster reachability table.

**Comment by helena** [comment]
@jroper what I meant was the flag in that samples hold in a more complex structure that would be in relation to the flag at given states of counts and times or whatever metadata is needed, slightly similar to that sample.

**Comment by helena** [comment]
What I like about the voting idea is that is one component of the swarm process which could be leveraged, aside from the purse algorithms.
